### PR TITLE
Feature lwdev 9226 cashout aggregation

### PR DESCRIPTION
**Pull request by OttoVT**

### CODE REVIEW

File: src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs [hunk 1]
```diff
@@ -138,7 +138,36 @@ public async Task Handle(SolarCashOutCompletedEvent evt, ICommandSender commandS
...
             commandSender.SendCommand(command, "operations");
         }
-        
+
+        #region CashourProcessorEvents
+
+        [UsedImplicitly]
+        public async Task Handle(CashoutsBatchCompletedEvent evt, ICommandSender commandSender)
+        {
+            _log.Info($"CashoutsBatchCompletedEvent for BIL received. BatchId [{evt.BatchId}]", evt);
```
Comment (some-dev) [inline src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs]: Looks like, exsessive logging, isn't it? Usually Cqrs-engine logs entry events/commands

File: src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs [hunk 2]
```diff
@@ -156,6 +185,69 @@ public async Task Handle(CashoutCompletedEvent evt, ICommandSender commandSender
...
+            }
+
+            foreach (var cashout in evt.Cashouts)
+            {
+                var command = new FailActivityCommand
+                {
+                    OperationId = cashout.OperationId,
+                    Output = new
+                    {
+                        ErrorCode = "BIL error",
```
Comment (some-dev) [inline src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs]: I think we should not use error Code as Magic string

File: src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs (hunk 1 above)
Comment (OttoVT) [reply src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs]: Just added what was before in operations service. 
Need one person to confirm that we should delete such log entrances

File: src/Lykke.Service.Operations/Modules/CqrsModule.cs [hunk 3]
```diff
@@ -150,7 +150,13 @@ protected override void Load(ContainerBuilder builder)
                             .ListeningEvents(typeof(Bitcoin.Contracts.Events.CashoutCompletedEvent)).From(BitcoinBoundedContext.Name).On("events")
                             .ListeningEvents(typeof(Job.EthereumCore.Contracts.Cqrs.Events.CashoutCompletedEvent)).From(EthereumBoundedContext.Name).On("events")
                             .ListeningEvents(typeof(SolarCashOutCompletedEvent)).From("solarcoin").On("events")
-                            .ListeningEvents(typeof(CashoutCompletedEvent)).From(BlockchainCashoutProcessorBoundedContext.Name).On("events")
+                            .ListeningEvents(
+                                typeof(CashoutCompletedEvent), 
+                                typeof(CashoutsBatchCompletedEvent),
+                                typeof(CrossClientCashoutCompletedEvent),
+                                typeof(CashoutFailedEvent),
+                                typeof(CashoutsBatchFailedEvent))
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.Operations/Modules/CqrsModule.cs]: Don't see that subscription to OperationExecutionFailedEvent was removed. CashoutFailedEvent and CashoutsBatchFailedEvent replace it

File: src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs (hunk 1 above)
Comment (KonstantinRyazantsev) [reply src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs]: We should delete this, if message logging is not disabled manually in CqrsEngine. Check it and remove such logging if Cqrs loggind is enabled

File: src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs [hunk 4]
```diff
@@ -138,7 +139,36 @@ public async Task Handle(SolarCashOutCompletedEvent evt, ICommandSender commandS
...
+        #region CashourProcessorEvents
+
+        [UsedImplicitly]
+        public async Task Handle(CashoutsBatchCompletedEvent evt, ICommandSender commandSender)
+        {
+            _log.Info($"CashoutsBatchCompletedEvent for BIL received. BatchId [{evt.BatchId}]", evt);
+
+            if (evt.Cashouts == null || evt.Cashouts.Length == 0)
+            {
+                _log.Warning($"Empty cashouts in batch. BatchId [{evt.BatchId}]", context: evt);
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs]: Better to throw an exception in such case. For instance, contract can be updated and message itself could be correct, but this handler should be altered to handle the message.

File: src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs [hunk 5]
```diff
@@ -156,6 +186,69 @@ public async Task Handle(CashoutCompletedEvent evt, ICommandSender commandSender
...
+        }
+
+        [UsedImplicitly]
+        public async Task Handle(CashoutsBatchFailedEvent evt, ICommandSender commandSender)
+        {
+            _log.Info($"CashoutsBatchFailedEvent for BIL received. BatchId [{evt.BatchId}]", evt);
+
+            if (evt.Cashouts == null || evt.Cashouts.Length == 0)
+            {
+                _log.Warning($"Empty cashouts in batch. BatchId [{evt.BatchId}]", context:evt);
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs]: As above

File: src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs [hunk 6]
```diff
@@ -156,6 +186,69 @@ public async Task Handle(CashoutCompletedEvent evt, ICommandSender commandSender
...
+
+                commandSender.SendCommand(command, "operations");
+            }
+        }
+
+        #endregion
+
+        [Obsolete]
+        [UsedImplicitly]
         public async Task Handle(OperationExecutionFailedEvent evt, ICommandSender commandSender)
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs]: This should be removed

File: src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs [hunk 7]
```diff
@@ -106,12 +103,35 @@ public async Task Handle(SolarCashOutCompletedEvent evt, ICommandSender commandS
...
+        #region CashourProcessorEvents
+
         [UsedImplicitly]
-        public async Task Handle(CashoutCompletedEvent evt, ICommandSender commandSender)
+        public async Task Handle(CashoutsBatchCompletedEvent evt, ICommandSender commandSender)
         {
-            _log.Info($"CashoutCompletedEvent for BIL received. Operation [{evt.OperationId}]", evt);
+            if (evt.Cashouts == null || evt.Cashouts.Length == 0)
+            {
+                throw new InvalidOperationException($"Empty cashouts in batch. BatchId [{evt.BatchId}]");
```
Comment (polonzer) [inline src/Lykke.Service.Operations/Workflow/Sagas/BlockchainCashoutSaga.cs]: Why throwing exception? Why this is a problem? What the service can do by retrying the event? The same applies to the code below.

**Pull request merged**